File: scripts/crawl-thebyte.js
const puppeteer = require('puppeteer-core');
const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

// Configuration
const BASE_URL = 'https://www.thebyte.com.cn';
const OUTPUT_DIR = '/Users/sphinx.shi/workspace/wiki/pyramid/raw/bookmarks/ebooks/thebyte';
const IMAGES_DIR = path.join(OUTPUT_DIR, 'images');
const MAX_IMAGES_PER_PAGE = 8;

// Pages to crawl
const pages = [
  '/intro.html',
  '/noun.html',
  '/architecture/summary.md',
  '/architecture/history.html',
  '/architecture/background.html',
  '/architecture/define-cloud-native.html',
  '/architecture/target.html',
  '/architecture/container.html',
  '/architecture/MicroService.html',
  '/architecture/ServiceMesh.html',
  '/architecture/Immutable.html',
  '/architecture/declarative-api.html',
  '/architecture/devops.html',
  '/architecture/arc.html',
  '/architecture/architect.html',
];

// Ensure directories exist
fs.mkdirSync(OUTPUT_DIR, { recursive: true });
fs.mkdirSync(IMAGES_DIR, { recursive: true });
fs.mkdirSync(path.join(OUTPUT_DIR, 'architecture'), { recursive: true });

async function downloadImage(imageUrl, pageName, index) {
  if (!imageUrl || imageUrl.startsWith('data:') || imageUrl.startsWith('//')) {
    return null;
  }

  // Convert protocol-relative URLs to https
  if (imageUrl.startsWith('//')) {
    imageUrl = 'https:' + imageUrl;
  }

  // Skip external URLs
  if (!imageUrl.startsWith(BASE_URL) && !imageUrl.startsWith('http')) {
    return null;
  }

  try {
    const urlObj = new URL(imageUrl);
    const pathname = urlObj.pathname;
    const ext = path.extname(pathname) || '.jpg';
    const filename = `${pageName}_img${index}${ext}`;
    const filepath = path.join(IMAGES_DIR, filename);

    const file = fs.createWriteStream(filepath);
    await new Promise((resolve, reject) => {
      const protocol = urlObj.protocol === 'https:' ? https : http;
      protocol.get(imageUrl, (response) => {
        if (response.statusCode === 200) {
          response.pipe(file);
          file.on('finish', () => {
            file.close();
            resolve();
          });
        } else {
          file.close();
          reject(new Error(`HTTP ${response.statusCode}`));
        }
      }).on('error', reject);
    });

    return filename;
  } catch (err) {
    console.log(`  Failed to download ${imageUrl}: ${err.message}`);
    return null;
  }
}

async function crawlPage(pagePath) {
  console.log(`\n=== Crawling: ${pagePath} ===`);

  const browser = await puppeteer.launch({
    executablePath: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  try {
    const page = await browser.newPage();
    await page.setViewport({ width: 1280, height: 800 });

    const url = BASE_URL + pagePath;
    console.log(`  Loading: ${url}`);

    await page.goto(url, { waitUntil: 'networkidle2', timeout: 30000 });

    // Wait for VuePress content to render
    await page.waitForSelector('.theme-default-content', { timeout: 10000 }).catch(() => {
      console.log('  Warning: .theme-default-content not found');
    });

    // Wait a bit more for dynamic content
    await new Promise(r => setTimeout(r, 2000));

    // Extract title
    const title = await page.evaluate(() => {
      const h1 = document.querySelector('h1');
      if (h1) {
        // Get text content only, removing any anchor symbols
        return h1.textContent.replace(/^#\s*/, '').trim();
      }
      return document.title;
    });
    console.log(`  Title: ${title}`);

    // Extract main content
    const content = await page.evaluate(() => {
      const contentEl = document.querySelector('.theme-default-content');
      if (!contentEl) return null;

      // Clone to avoid modifying original
      const clone = contentEl.cloneNode(true);

      // Remove script and style elements
      clone.querySelectorAll('script, style, .sidebar, .navbar, .page-nav, .page-meta, .footer, .qrcode, .giscus-wrapper, .page-info').forEach(el => el.remove());

      return clone.innerHTML;
    });

    if (!content) {
      console.log('  Warning: No content extracted');
    }

    // Extract images (max 8)
    const maxImages = MAX_IMAGES_PER_PAGE;
    const images = await page.evaluate((max) => {
      const contentEl = document.querySelector('.theme-default-content');
      if (!contentEl) return [];

      const imgElements = contentEl.querySelectorAll('img');
      return Array.from(imgElements).slice(0, max).map(img => img.src);
    }, maxImages);

    console.log(`  Found ${images.length} images`);

    // Download images
    const pageName = pagePath.replace(/^\//, '').replace(/\//g, '_').replace(/\.html?$/, '').replace(/\.md$/, '');
    const downloadedImages = [];

    for (let i = 0; i < images.length; i++) {
      const img = await downloadImage(images[i], pageName, i);
      if (img) {
        downloadedImages.push(img);
        console.log(`  Downloaded: ${img}`);
      }
    }

    // Replace image URLs in content with local filenames
    let processedContent = content;
    for (let i = 0; i < downloadedImages.length; i++) {
      if (images[i]) {
        processedContent = processedContent.replace(new RegExp(images[i].replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g'), `images/${downloadedImages[i]}`);
      }
    }

    // Convert HTML to Markdown
    const markdown = htmlToMarkdown(processedContent, title);

    // Determine output file path
    let outputPath;
    if (pagePath.startsWith('/architecture/')) {
      outputPath = path.join(OUTPUT_DIR, 'architecture', path.basename(pagePath));
    } else {
      outputPath = path.join(OUTPUT_DIR, path.basename(pagePath));
    }

    // Change extension to .md
    outputPath = outputPath.replace(/\.html$/, '.md');

    fs.writeFileSync(outputPath, markdown);
    console.log(`  Saved: ${outputPath}`);

    return { title, path: pagePath, outputPath };

  } finally {
    await browser.close();
  }
}

function htmlToMarkdown(html, title) {
  if (!html) return '';

  // First, remove header anchor links like <a class="header-anchor" ...>#</a>
  html = html.replace(/<a[^>]*class="header-anchor"[^>]*>[^<]*<\/a>/gi, '');

  // Simple HTML to Markdown conversion
  let md = html
    // Remove VuePress specific elements - custom containers
    .replace(/<div class="custom-container[^"]*">/g, '\n:::')
    .replace(/<\/div>/g, '\n:::\n')
    // Headers - simple approach, just strip all tags from content
    .replace(/<h1[^>]*>([\s\S]*?)<\/h1>/gi, (_, c) => '# ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    .replace(/<h2[^>]*>([\s\S]*?)<\/h2>/gi, (_, c) => '## ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    .replace(/<h3[^>]*>([\s\S]*?)<\/h3>/gi, (_, c) => '### ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    .replace(/<h4[^>]*>([\s\S]*?)<\/h4>/gi, (_, c) => '#### ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    .replace(/<h5[^>]*>([\s\S]*?)<\/h5>/gi, (_, c) => '##### ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    .replace(/<h6[^>]*>([\s\S]*?)<\/h6>/gi, (_, c) => '###### ' + c.replace(/<[^>]+>/g, '').trim() + '\n\n')
    // Links
    .replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, '[$2]($1)')
    // Bold and italic
    .replace(/<strong[^>]*>([\s\S]*?)<\/strong>/gi, '**$1**')
    .replace(/<b[^>]*>([\s\S]*?)<\/b>/gi, '**$1**')
    .replace(/<em[^>]*>([\s\S]*?)<\/em>/gi, '*$1*')
    .replace(/<i[^>]*>([\s\S]*?)<\/i>/gi, '*$1*')
    // Code
    .replace(/<code[^>]*>([\s\S]*?)<\/code>/gi, '`$1`')
    .replace(/<pre[^>]*>([\s\S]*?)<\/pre>/gi, '\n```\n$1\n```\n')
    // Lists
    .replace(/<ul[^>]*>/gi, '\n')
    .replace(/<\/ul>/gi, '\n')
    .replace(/<ol[^>]*>/gi, '\n')
    .replace(/<\/ol>/gi, '\n')
    .replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, '- $1\n')
    // Paragraphs and line breaks
    .replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, '\n$1\n\n')
    .replace(/<br\s*\/?>/gi, '\n')
    // Images (will be replaced with local paths by caller)
    .replace(/<img[^>]*src="([^"]*)"[^>]*>/gi, '![]($1)')
    .replace(/<img[^>]*alt="([^"]*)"[^>]*>/gi, '![]($1)')
    // Divs and spans - remove
    .replace(/<div[^>]*>/gi, '\n')
    .replace(/<\/div>/gi, '\n')
    .replace(/<span[^>]*>([\s\S]*?)<\/span>/gi, '$1')
    // Clean up empty tags
    .replace(/<[^>]+>:/gi, ':')
    .replace(/<[^>]+>/gi, '')
    // Clean up extra whitespace
    .replace(/\n{3,}/g, '\n\n')
    .replace(/^\s+/gm, '')
    .trim();

  return md;
}

async function main() {
  console.log('Starting crawl of thebyte.com.cn...');
  console.log(`Output directory: ${OUTPUT_DIR}`);

  const results = [];

  for (const pagePath of pages) {
    try {
      const result = await crawlPage(pagePath);
      results.push(result);
    } catch (err) {
      console.error(`  Error crawling ${pagePath}: ${err.message}`);
    }
  }

  console.log('\n=== Crawl Complete ===');
  console.log(`Processed ${results.length} pages`);
}

main().catch(console.error);

File: scripts/crawl_thebyte.py
#!/usr/bin/env python3
"""
Crawl pages from thebyte.com.cn and convert to markdown.
"""

import os
import re
import html
from pathlib import Path
from urllib.parse import urljoin, urlparse
import subprocess
import time

BASE_URL = "https://www.thebyte.com.cn"
OUTPUT_DIR = Path("/Users/sphinx.shi/workspace/wiki/pyramid/raw/bookmarks/ebooks/thebyte")
IMAGES_DIR = OUTPUT_DIR / "images"
IMAGES_DIR.mkdir(parents=True, exist_ok=True)

CHAPTERS = {
    "consensus": [
        "/consensus/summary.html",
        "/consensus/consensus.html",
        "/consensus/Replicated-State-Machine.html",
        "/consensus/Paxos.html",
        "/consensus/Paxos-history.html",
        "/consensus/Basic-Paxos.html",
        "/consensus/Raft.html",
        "/consensus/raft-leader-election.html",
        "/consensus/raft-log-replication.html",
        "/consensus/raft-ConfChange.html",
    ],
    "container": [
        "/container/summary.html",
        "/container/borg-omega-k8s.html",
        "/container/orchestration.html",
        "/container/image.html",
        "/container/CRI.html",
        "/container/storage.html",
        "/container/container-network.html",
        "/container/Resource-scheduling.html",
        "/container/resource.html",
        "/container/Extended-Resource.html",
        "/container/kube-scheduler.html",
        "/container/auto-scaling.html",
    ],
}

def download_image(img_url, img_name, page_name):
    """Download an image and return the local path."""
    if not img_url or img_url.startswith('data:'):
        return None

    # Skip external URLs
    parsed = urlparse(img_url)
    if parsed.netloc and parsed.netloc != 'thebyte.com.cn' and parsed.netloc != 'www.thebyte.com.cn':
        return None

    local_name = f"{page_name}_{img_name}"
    local_path = IMAGES_DIR / local_name

    if local_path.exists():
        return local_name

    # Download with curl
    if img_url.startswith('/'):
        full_url = BASE_URL + img_url
    else:
        full_url = img_url

    try:
        subprocess.run(
            ["curl", "-s", "-L", "-o", str(local_path), full_url],
            check=True,
            timeout=30
        )
        if local_path.exists() and local_path.stat().st_size > 0:
            return local_name
        else:
            return None
    except Exception as e:
        print(f"  Failed to download {full_url}: {e}")
        return None


def extract_content(html_content, page_path):
    """Extract article content from HTML."""
    # Find theme-default-content div
    match = re.search(r'<div class="theme-default-content">(.*?)</div>\s*<!--\[-->\s*<!--\[-->\s*<div class="page-info"', html_content, re.DOTALL)
    if not match:
        # Try alternative pattern
        match = re.search(r'<div class="theme-default-content">(.*?)<div class="page-info"', html_content, re.DOTALL)

    if not match:
        print(f"  Could not find theme-default-content for {page_path}")
        return None

    return match.group(1)


def html_to_markdown(html_content, page_path, page_name):
    """Convert HTML content to markdown."""
    content = html_content

    # Track images for download
    images_to_download = []
    img_counter = 0

    # Extract and replace images
    def replace_img(match):
        nonlocal img_counter
        img_src = match.group(1)
        img_alt = match.group(2) if match.group(2) else f"image"

        if img_counter >= 8:
            return ""

        # Create a simple name for the image
        img_name = f"img{img_counter:02d}.png"
        img_counter += 1

        images_to_download.append((img_src, img_name))
        return f"\n![{img_alt}](../images/{page_name}_{img_name})\n"

    # Replace images
    content = re.sub(r'<img[^>]+src=["\']([^"\']+)["\'][^>]*alt=["\']([^"\']*)["\'][^>]*>', replace_img, content)
    content = re.sub(r'<img[^>]+src=["\']([^"\']+)["\'][^>]*>', replace_img, content)

    # Download images
    for img_src, img_name in images_to_download[:8]:
        download_image(img_src, img_name, page_name)

    # Remove VuePress comments
    content = re.sub(r'<!--\[-->|<!--\]-->|<!\[CDATA\[|\]\]>', '', content)

    # Remove empty comments
    content = re.sub(r'<!--.*?-->', '', content, flags=re.DOTALL)

    # Remove script tags
    content = re.sub(r'<script[^>]*>.*?</script>', '', content, flags=re.DOTALL)

    # Remove style tags
    content = re.sub(r'<style[^>]*>.*?</style>', '', content, flags=re.DOTALL)

    # Remove header anchors
    content = re.sub(r'<a class="header-anchor"[^>]*>[^<]*</a>', '', content)

    # Remove giscus wrapper
    content = re.sub(r'<div class="giscus-wrapper[^"]*"[^>]*>.*?</div>', '', content, flags=re.DOTALL)

    # Remove qrcode div
    content = re.sub(r'<div class="qrcode"[^>]*>.*?</div>', '', content, flags=re.DOTALL)

    # Remove page-nav
    content = re.sub(r'<div class="page-nav[^"]*"[^>]*>.*?</div>', '', content, flags=re.DOTALL)

    # Remove page-meta footer
    content = re.sub(r'<footer class="page-meta[^"]*"[^>]*>.*?</footer>', '', content, flags=re.DOTALL)

    # Remove data-v attributes
    content = re.sub(r'\s*data-v-[a-z0-9]+="[^"]*"', '', content)
    content = re.sub(r'\s*data-v-[a-z0-9]+', '', content)

    # Extract title
    title_match = re.search(r'<h1[^>]*id="([^"]*)"[^>]*>([^<]*)</h1>', content)
    title = title_match.group(2) if title_match else os.path.basename(page_path).replace('.html', '')

    # Remove h1 since we'll add it in frontmatter
    content = re.sub(r'<h1[^>]*>.*?</h1>', '', content, flags=re.DOTALL)

    # Process custom containers (tip, warning, danger, info)
    def replace_container(match):
        container_type = match.group(1)
        inner = match.group(2)
        # Extract title if exists
        title_match = re.search(r'<p class="custom-container-title"><a></a></p>\s*<p>([^<]*)</p>', inner)
        title = title_match.group(1) if title_match else ""
        # Get paragraph content
        p_match = re.search(r'<p>([^<].*?)</p>', inner, re.DOTALL)
        text = p_match.group(1) if p_match else inner
        text = re.sub(r'<[^>]+>', '', text).strip()
        if title:
            return f"\n> **{title}** {text}\n"
        return f"\n> {text}\n"

    content = re.sub(r'<div class="custom-container ([^"]+)">\s*<p class="custom-container-title"><a></a></p>(.*?)</div>', replace_container, content, flags=re.DOTALL)

    # Process div custom-container (center, right)
    content = re.sub(r'<div class="custom-container center">\s*<p>(.*?)</p>\s*</div>', r'\n<div align="center">\1</div>\n', content, flags=re.DOTALL)
    content = re.sub(r'<div class="custom-container right">\s*<p>(.*?)</p>\s*</div>', r'\n<div align="right">\1</div>\n', content, flags=re.DOTALL)

    # Remove remaining divs
    content = re.sub(r'</?div[^>]*>', '', content)

    # Convert headers
    content = re.sub(r'<h(\d)[^>]*>([^<]*)</h\1>', r'\n## \2\n', content)

    # Convert paragraphs
    content = re.sub(r'<p[^>]*>(.*?)</p>', r'\n\1\n', content, flags=re.DOTALL)

    # Convert line breaks
    content = re.sub(r'<br\s*/?>','\n', content)

    # Convert strong/bold
    content = re.sub(r'<strong[^>]*>([^<]*)</strong>', r'**\1**', content)
    content = re.sub(r'<b[^>]*>([^<]*)</b>', r'**\1**', content)

    # Convert emphasis
    content = re.sub(r'<em[^>]*>([^<]*)</em>', r'*\1*', content)

    # Convert links
    content = re.sub(r'<a[^>]+href=["\']([^"\']+)["\'][^>]*>([^<]*)</a>', r'[\2](\1)', content)

    # Remove remaining HTML tags
    content = re.sub(r'<[^>]+>', '', content)

    # Decode HTML entities
    content = html.unescape(content)

    # Clean up whitespace
    content = re.sub(r'\n{3,}', '\n\n', content)
    content = content.strip()

    return title, content


def process_page(url_path):
    """Process a single page."""
    url = BASE_URL + url_path
    print(f"Processing: {url_path}")

    # Determine output path
    if url_path.startswith('/'):
        rel_path = url_path.lstrip('/')
    else:
        rel_path = url_path

    # Convert .html to .md
    if rel_path.endswith('.html'):
        md_path = rel_path.replace('.html', '.md')
    else:
        md_path = rel_path + '.md'

    output_file = OUTPUT_DIR / md_path

    # Create parent directories
    output_file.parent.mkdir(parents=True, exist_ok=True)

    # Fetch HTML
    result = subprocess.run(
        ["curl", "-s", "-L", url],
        capture_output=True,
        text=True,
        timeout=60
    )

    if result.returncode != 0:
        print(f"  Failed to fetch {url}")
        return False

    html_content = result.stdout

    # Extract title
    title_match = re.search(r'<title>([^<]*)</title>', html_content)
    page_title = title_match.group(1) if title_match else os.path.basename(url_path)

    # Extract content
    content = extract_content(html_content, url_path)
    if not content:
        print(f"  Could not extract content from {url_path}")
        return False

    # Convert to markdown
    article_title, md_content = html_to_markdown(content, url_path, md_path.replace('/', '_').replace('.md', ''))

    # Build full markdown
    frontmatter = f"""---
title: "{article_title}"
source: thebyte
url: "{url}"
---

"""
    full_md = frontmatter + md_content

    # Write output
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(full_md)

    print(f"  Saved: {output_file}")
    return True


def main():
    total = 0
    success = 0

    for chapter, pages in CHAPTERS.items():
        print(f"\n=== Processing {chapter} ===")
        for page in pages:
            total += 1
            if process_page(page):
                success += 1
            time.sleep(0.5)  # Be polite

    print(f"\n\nDone: {success}/{total} pages processed successfully")


if __name__ == "__main__":
    main()